File: Elego/node.py
import numpy as np
from move import Move
class Node():
    def __init__(self,gamestate,parent,last_move,priors,value):
        self.gamestate = gamestate
        self.parent = parent
        self.last_move = last_move
        self.visits = 1
        self.value = value
        self.size = gamestate.board.size
        self.is_terminal = gamestate.is_over()
        self.virtual_losses = 0

        self.branch_values = np.zeros((self.size*self.size),dtype=np.float32)
        self.branch_priors = priors
        self.branch_visits = np.zeros((self.size*self.size),dtype=np.float32)

        self.children = {}
        self.branches = set()

        #create set of legal branches
        for mv_index in range(self.size*self.size):    
            mv_x = mv_index%self.size + 1
            mv_y = mv_index//self.size+1
            move = Move(gamestate.turn,mv_x,mv_y)
            if(gamestate.is_legal_move(move)):
                self.branches.add(mv_index)

    def get_total_value(self):
        return self.values

    def set_total_value(self,values):
        self.values = values

    def get_visit_count(self):
        return self.branch_visits

    def set_visit_count(self,visits):
        self.branch_visits = visits

    def select_branch(self):
        moves = np.argsort(self.score_branch())[::-1] #sort by descending order
        for move in moves:
            if move in self.branches:
                return move

    def select_move(self):
        return np.argmax(self.branch_visits)

    def add_child(self,move,child):
        self.children[move] = child
        #self.visits += 1
        #self.branch_visits[move] += 1
        #self.branch_values[move] -= child.value

    def has_child(self,move):
        return move in self.children

    def get_child(self,move):
        return self.children[move]
            
    def score_branch(self):
        return self.branch_Q() + self.branch_U()

    def branch_Q(self):
        return np.divide(self.branch_values, self.branch_visits,
        out=np.zeros_like(self.branch_values), where=self.branch_visits!=0)

    def branch_U(self):
        c=1.4
        return c*self.branch_priors*np.sqrt(self.visits)/(self.branch_visits+1)

    def update_wins(self,value):
        if not self.parent is None:
            parent = self.parent
            parent.visits = parent.visits + 1
            parent.branch_visits[self.last_move] += 1
            parent.branch_values[self.last_move] += value
            parent.update_wins(-value)

    def add_virtual_loss(self):
        self.virtual_losses += 1
        self.visits += 1
        self.value -= 1
        if not self.parent is None:
            self.parent.branch_values[self.last_move] -= 1
            self.parent.branch_visits[self.last_move] += 1
            self.parent.add_virtual_loss()

    def undo_virtual_losses(self):
        self.visits -= self.virtual_losses
        self.value += self.virtual_losses
        if not self.parent is None:
            self.parent.branch_values[self.last_move] += self.virtual_losses
            self.parent.branch_visits[self.last_move] -= self.virtual_losses
            self.parent.undo_virtual_losses()
        self.virtual_losses = 0

    def print(self):
        self.gamestate.print_state()
        print("Number of visits:",self.visits)
        print("Value:",self.value)
        print("Branch visits:",self.branch_visits)
        print("Branch values:",self.branch_values)
        print("Branch priors:",self.branch_priors)
        if not self.branch_priors is None:
            print("Explore score:",(self.score_branch()))

File: SelfplayEngine/agents/node.cc
#include "node.h"
#include <iostream>
#include <cmath>

double Node::branch_Q(Point move){
    Branch branch = branches.at(move);
    if(branch.visits == 0){
        return 0;
    }
    return branch.total_value/branch.visits;
}
double Node::branch_U(Point move){
    Branch branch = branches.at(move);
    double c = 1.1;
    return c*branch.prior*sqrt(visits)/(1+branch.visits);
}
double Node::branch_score(Point move){
    return branch_Q(move) + branch_U(move);
}


void Node::add_child(Point move, std::shared_ptr<Node> child){
    children.insert({move,child});
}

void Node::update_branch(Point move, double value){
    Branch branch = branches.at(move);
    branch.visits++;
    branch.total_value += value;
    branches.at(move) = branch;
    visits++;
    if (parent != nullptr){
        parent->update_branch(last_move,-value);
    }
}

Point Node::select_branch(){
    double max_score = -10.0;
    Point best_move({0,0});
    for (auto branch: branches){
        double score = branch_score(branch.first);
        if (score > max_score){
            max_score = score;
            best_move = branch.first;
        }
    }
    return best_move;
}

Point Node::select_move(){
    int max_visits = -1.0;
    Point best_move({-1,-1});
    for (auto branch : branches){
        int branch_visits = branch.second.visits;
        if (branch_visits > max_visits){
            max_visits = branch_visits;
            best_move = branch.first;
        }
    }
    return best_move;
}

void Node::print_scores(){
    for (auto branch: branches){
        std::cout<<branch.first.first<<","<<branch.first.second<<": "
        <<branch_Q(branch.first)<<" + "<<branch_U(branch.first)<<" = "<<branch_score(branch.first)<<"\n";
    }
}
void Node::print_visits(){
    for (auto branch: branches){
        std::cout<<branch.first.first<<","<<branch.first.second<<": "<<branch.second.visits<<"\n";
    }
}